go-runtime/schema/analyzers/verb.go

```diff
@@ -0,0 +1,188 @@
+			Export:   isExported,
+			Name:     strcase.ToLowerCamel(node.Name.Name),
+			Request:  reqV,
+			Response: resV,
+			Metadata: metadata,
+		}
+		nn[verb] = node.Name.Name
+		verbs = append(verbs, verb)
+	})
+
+	return newExtractorResult(pass, verbs, nn), nil
+}
+
+func extractVerbMetadata(pass *analysis.Pass, node ast.Node, directives []directive) []schema.Metadata {
+	metadata := []schema.Metadata{}
+	for _, dir := range directives {
+		switch dir := dir.(type) {
+		case *directiveIngress:
+			typ := dir.Type
+			if typ == "" {
+				typ = "http"
+			}
+			metadata = append(metadata, &schema.MetadataIngress{
+				Pos:    dir.Pos,
+				Type:   typ,
+				Method: dir.Method,
+				Path:   dir.Path,
+			})
+		case *directiveCronJob:
+			metadata = append(metadata, &schema.MetadataCronJob{
+				Pos:  dir.Pos,
+				Cron: dir.Cron,
+			})
+		case *directiveRetry:
+			metadata = append(metadata, &schema.MetadataRetry{
+				Pos:        dir.Pos,
+				Count:      dir.Count,
+				MinBackoff: dir.MinBackoff,
+				MaxBackoff: dir.MaxBackoff,
+			})
+		case *directiveSubscriber:
+			metadata = append(metadata, &schema.MetadataSubscriber{
+				Pos:  dir.Pos,
+				Name: dir.Name,
+			})
+		case *directiveVerb:
+		case *directiveData, *directiveEnum, *directiveTypeAlias, *directiveExport:
+			pass.Report(errorf(node, "unexpected directive %T", dir))
+		}
+	}
+	return metadata
+}
+
+func checkSignature(pass *analysis.Pass, node *ast.FuncDecl, sig *types.Signature) (req, resp optional.Option[*types.Var]) {
+	params := sig.Params()
+	results := sig.Results()
+
+	if params.Len() > 2 {
+		pass.Report(errorf(node, "must have at most two parameters (context.Context, struct)"))
+	}
+
+	loaded := pass.ResultOf[Initializer].(initializeResult) //nolint:forcetypeassert
+	if params.Len() == 0 {
+		pass.Report(errorf(node, "first parameter must be context.Context"))
+	} else if !loaded.isFtlContextType(params.At(0).Type()) {
+		pass.Report(tokenErrorf(pass.Fset, params.At(0).Pos(), params.At(0).Name(), "first parameter must be of type context.Context but is %s", params.At(0).Type()))
+	}
+
+	if params.Len() == 2 {
+		if !isType[*types.Struct](params.At(1).Type()) {
+			pass.Report(tokenErrorf(pass.Fset, params.At(1).Pos(), params.At(1).Name(), "second parameter must be a struct but is %s", params.At(1).Type()))
+		}
+		if params.At(1).Type().String() == ftlUnitTypePath {
+			pass.Report(tokenErrorf(pass.Fset, params.At(1).Pos(), params.At(1).Name(), "second parameter must not be ftl.Unit"))
+		}
+
+		req = optional.Some(params.At(1))
+	}
+
+	if results.Len() > 2 {
+		pass.Report(errorf(node, "must have at most two results (struct, error)"))
+	}
+	if results.Len() == 0 {
+		pass.Report(errorf(node, "must at least return an error"))
+	} else if !loaded.isFtlErrorType(results.At(results.Len() - 1).Type()) {
+		pass.Report(tokenErrorf(pass.Fset, results.At(results.Len()-1).Pos(), results.At(results.Len()-1).Name(), "must return an error but is %s", results.At(0).Type()))
```

Comment:
In this case, when returning multiple values, should we specify in the error message that _last_ value in the return statement must be `error`, in case they have something like `func(context.Context) (error, struct)`?

Reply:
I think that's what this is already doing isn't it? `results.At(results.Len()-1)` is the last out value.

Reply:
Ah I just meant clarifying the error message. The logic itself looks good yeah